### ISSUE TITLE
Should `_forms` be required?

### ISSUE THREAD

**Issue by pezra** [Dwolla/hal-forms#24]
from @travi in another thread:

[`_forms` optionality] might be worth clarifying more, but my understanding is that is is optional only in the case where the media-type, with the proper profile, is requested, but when a form representation of the requested resource has not been defined. in other words, it MUST be included if available, but SHOULD be omitted if not present. i would not consider its inclusion to be optional, or potentially even allowed if not requested in the Accept header. @jareddellitt is this the intended interpretation?

**Comment by pezra** [comment]
Adopting this interpretation would be detrimental. Having alternate representations of the same resource, that are used by the same consumer, would make consumer implementation unnecessarily difficult. Consumers would have to decide on each request whether or not to add the profile attribute to the accept header. They would have to manage multiple cache entries per resource. They would be required to make multiple HTTP requests to get all the information about a single resource. For that complication, they would gain nothing of value. 

If we treat forms as supplemental data clients can be dramatically simplified and we make them more discoverable. Consumers can have just one accept header. They can know that when the fetch a resource they know all its affordances. They only need one cache entry for each resource. Developers who don't know about forms can see the forms, and profile information, in their exploratory requests and they go an learn more.

All of the following flows should be legal:
![flow-diagram](https://user-images.githubusercontent.com/4122/40377279-bdf1ab3c-5dad-11e8-90b2-727940dcc201.png)

**Comment by jareddellitt** [comment]
Looking at the `profile` [spec](https://tools.ietf.org/html/rfc6906#section-3) a bit closer, it reads:

> A profile MUST NOT change the semantics of the
   resource representation when processed without profile knowledge, so
   that clients both with and without knowledge of a profiled resource
   can safely use the same representation.  While this specification
   associates profiles with resource representations, creators and users
   of profiles MAY define and manage them in a way that allows them to
   be used across media types; thus, they could be associated with a
   resource, independent of their representations (i.e., using the same
   profile URI for different media types).

I think this is getting at what @pezra is talking about, which means that `_forms` should be supplemental and we've misinterpreted what the `profile` represents (i.e., not a different representation of the resource). This means that the flows pictured above, would be valid.

This doesn't answer the question about whether or not `_forms` are returned when a client doesn't ask for them, which may just be something that the implementer can choose to support if they wish, but the consumer can just ignore.

**Comment by pezra** [comment]
> question about whether or not _forms are returned when a client doesn't ask for them

yeah, that is a bit of a grey area. i propose we say something like producers MAY include the`_forms` property even if the consumer doesn't request.

**Comment by travi** [comment]
> I think this is getting at what @pezra is talking about, which means that _forms should be supplemental and we've misinterpreted what the profile represents (i.e., not a different representation of the resource). This means that the flows pictured above, would be valid.

that makes sense based on that 

> i propose we say something like producers MAY include the_forms property even if the consumer doesn't request

i'm still a little on the fence about what is appropriate messaging here. this seems very much up to the implementation choices made by the implementors of the provider, but it seems outside the scope of this spec (assuming knowledge of profiles in general, which we are admittedly showing some confusion around). 

the information presented under `_forms` seems like it should be opt-in. as a consumer, i'd like to at least be given the choice around whether i need that information as part of the response. if i am just rendering a read-only list of resources, i would like to omit the `_forms` information from needing to be transferred over the wire (and potentially calculated by the provider). from that perspective, i think it is valuable to suggest that a provider SHOULD NOT include it without it being asked for (which is not MUST NOT), but again that is more a decision around flexibility for the provider implementers to decide. 

i think this part comes down to how prescriptive this spec should be around more general hal practices...

**Comment by pezra** [comment]
we could, of course, remain silent on this issue and leave it entirely up to the community.
